flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/polling/PollingRecordPublisher.java

```diff
@@ -0,0 +1,162 @@
+	}
+
+	public RecordPublisherRunResult run(final StartingPosition startingPosition, final Consumer<RecordBatch> consumer, int maxNumberOfRecords) throws InterruptedException {
+		if (nextShardItr == null) {
+			nextShardItr = getShardIterator(startingPosition);
+		}
+
+		if (nextShardItr == null) {
+			return COMPLETE;
+		}
+
+		metricsReporter.setMaxNumberOfRecordsPerFetch(maxNumberOfRecords);
+
+		GetRecordsResult result = getRecords(nextShardItr, startingPosition, maxNumberOfRecords);
+
+		consumer.accept(new RecordBatch(result.getRecords(), subscribedShard, result.getMillisBehindLatest()));
+
+		nextShardItr = result.getNextShardIterator();
+		return nextShardItr == null ? COMPLETE : INCOMPLETE;
+	}
+
+	/**
+	 * Calls {@link KinesisProxyInterface#getRecords(String, int)}, while also handling unexpected
+	 * AWS {@link ExpiredIteratorException}s to assure that we get results and don't just fail on
+	 * such occasions. The returned shard iterator within the successful {@link GetRecordsResult} should
+	 * be used for the next call to this method.
+	 *
+	 * <p>Note: it is important that this method is not called again before all the records from the last result have been
+	 * fully collected with {@code ShardConsumer#deserializeRecordForCollectionAndUpdateState(UserRecord)}, otherwise
+	 * {@code ShardConsumer#lastSequenceNum} may refer to a sub-record in the middle of an aggregated record, leading to
+	 * incorrect shard iteration if the iterator had to be refreshed.
+	 *
+	 * @param shardItr shard iterator to use
+	 * @param startingPosition the position in the stream in which to consume from
+	 * @param maxNumberOfRecords the maximum number of records to fetch for this getRecords attempt
+	 * @return get records result
+	 */
+	private GetRecordsResult getRecords(String shardItr, final StartingPosition startingPosition, int maxNumberOfRecords) throws InterruptedException {
+		GetRecordsResult getRecordsResult = null;
+		while (getRecordsResult == null) {
+			try {
+				getRecordsResult = kinesisProxy.getRecords(shardItr, maxNumberOfRecords);
+			} catch (ExpiredIteratorException eiEx) {
+				LOG.warn("Encountered an unexpected expired iterator {} for shard {};" +
+					" refreshing the iterator ...", shardItr, subscribedShard);
+
+				shardItr = getShardIterator(startingPosition);
+
+				// sleep for the fetch interval before the next getRecords attempt with the refreshed iterator
+				if (expiredIteratorBackoffMillis != 0) {
+					Thread.sleep(expiredIteratorBackoffMillis);
+				}
+			}
+		}
+		return getRecordsResult;
+	}
+
+	/**
+	 * Returns a shard iterator for the given {@link SequenceNumber}.
+	 *
+	 * @return shard iterator
+	 */
+	@Nullable
+	private String getShardIterator(final StartingPosition startingPosition) throws InterruptedException {
+		if (startingPosition.getShardIteratorType() == LATEST && subscribedShard.isClosed()) {
+			return null;
+		}
+
+		return kinesisProxy.getShardIterator(
```

Comment:
For timestamp the AWS client throws an exception if `time > now`. If `time <= now` it returns `null` if the shard is complete, otherwise the iterator.

If a sequence number is non-existent then the service throws a 400 client error and the application terminates. If the sequence number is invalid then the AWS client throws an exception.